### ISSUE TITLE
Several major issues with changed sources detection.

### ISSUE THREAD

**Issue by Sap333** [eth-brownie/brownie#1231]
### Environment information

* `brownie` Version: 1.15.6
* Python Version: 3.9.6
* OS: Linux

### What was wrong?

Brownie recompiles contracts/interfaces even if they didn't change. Actually it recompiles all contracts every time. The bug depends (but not only) on required/downloaded compiler version and compiler options.

### How can it be fixed?

There are still several major issues with detecting of changed sources.
There is a mistake in _brownie/project/main.py_ line 264 in method _Project_.__get_changed_contracts_:
`for name in [k for k, v in new_hashes.items() if compiled_hashes.get(k, None) != v]:`
It must be "==" there:
`for name in [k for k, v in new_hashes.items() if compiled_hashes.get(k, None) == v]:`

Another  thing is that method _Project_.__get_changed_contracts()_ always says that compiler settings has been changed. Something is deeply wrong with saving/comparing of the compiler settings.

And one more a mistake in line 276 in method _Project_.__get_changed_contracts_:
It says:
`changed_set: Set = set(self._sources.get_source_path(i) for i in contracts)`
which is obviously wrong because it returns the actual set of contracts, not the list of changed contracts/interfaces.
It seems like there must be something like:
`changed_set: Set = set(self._sources.get_source_path(i) for i in self._build.items())`

But actually the whole  _Project_.__get_changed_contracts()_  is a mess and must be checked/rewritten.

 The issue is related with #956 and #1194.

**Comment by iamdefinitelyahuman** [comment]
Agreed, that section of the code could definitely use a refactor.  Will happily receive a pull request.

**Comment by maxsiz** [comment]
I think  this issue may relate to #1221

**Comment by princesscrypto1337** [comment]
I had been running into the same issue. After digging around the source code, I found the issue is caused by defining the same interface in different files. Brownie doesn't emit an error when this happens. During compilation, you end up comparing the source code of one occurrence of the interface against another in `_compare_build_json`, which then flags the interface for recompilation along with all contracts that depend on it. If I just make sure I only define my interface once in some file and import from it, then the compilation detection works fine. Alternatively, one could just rename the interface uniquely in different files.

So perhaps a good solution on Brownie's end is to just emit an error message when multiple occurrences of the same interface are found, like it does for contracts.
